examples/datalog/stratifier.flix

```diff
@@ -0,0 +1,102 @@
+namespace Stratifier {
+    use Int32.max;
+    use List.{map, flatMap, partition};
+    use Array.toMap;
+
+    opaque type Program = List[Constraint]
+    opaque type Constraint = (HeadAtom, List[BodyAtom])
+    opaque type Atom = PredicateSymbol
+    opaque type HeadAtom = Atom
+    enum BodyAtom {
+        case Positive(Atom)
+        case Negative(Atom)
+    }
+    opaque type PredicateSymbol with Boxable, ToString, Eq, Order = String
+
+    // A :- B leads to (A, true, B)
+    type alias PrecedenceEdge = (PredicateSymbol, Bool, PredicateSymbol)
+    type alias PrecedenceGraph = List[PrecedenceEdge]
+
+    def convertProgram(p: Program): PrecedenceGraph =
+        let Program(c) = p;
+        flatMap(convertConstraint, c)
+
+    def convertConstraint(c: Constraint): PrecedenceGraph =
+        let Constraint(HeadAtom(Atom(a0)), b) = c;
+        let unfoldBodyAtom = ba -> match ba {
+            case Positive(Atom(ps)) => (true, ps),
+            case Negative(Atom(ps)) => (false, ps)
+        };
+        b |> map(unfoldBodyAtom) |> map(pna -> match pna {case (pn, a) => (a0, pn, a)})
+
+    pub def ullman(p: Program, numberOfPredicates: Int): Option[Map[PredicateSymbol, Int]] & Impure =
+        let pg: PrecedenceGraph = convertProgram(p);
+        let (pos, neg) = partition(e -> match e {case (_, b, _) => b}, pg);
+        let removeBool = triple -> match triple {case (a, _, b) => (a, b)};
+        let facts = project map(removeBool, pos), map(removeBool, neg) into PositiveDependencyEdge, NegativeDependencyEdge;
+        let rules = #{
+            Stratum(pd; 1) :- PositiveDependencyEdge(pd, _).
+            Stratum(pd; 1) :- PositiveDependencyEdge(_, pd).
+            Stratum(pd; 1) :- NegativeDependencyEdge(pd, _).
+            Stratum(pd; 1) :- NegativeDependencyEdge(_, pd).
```

Comment:
Is it necessary to set them to 1. If not they should be 0 and everything starts from there.

Reply:
I would suspect the above 4 lines can be omitted.

Reply:
The other stratum rules are recursive, but I've changed the start to 0